File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-asset-5-4/Underlyer.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class Underlyer : ISerialized
    {
        public Underlyer(XmlNode xmlNode)
        {
            XmlNode singleUnderlyerNode = xmlNode.SelectSingleNode("singleUnderlyer");
            
            if (singleUnderlyerNode != null)
            {
                if (singleUnderlyerNode.Attributes["href"] != null || singleUnderlyerNode.Attributes["id"] != null) 
                {
                    if (singleUnderlyerNode.Attributes["id"] != null) 
                    {
                        singleUnderlyerIDRef_ = singleUnderlyerNode.Attributes["id"].Value;
                        SingleUnderlyer ob = new SingleUnderlyer(singleUnderlyerNode);
                        IDManager.SetID(singleUnderlyerIDRef_, ob);
                    }
                    else if (singleUnderlyerNode.Attributes["href"] != null)
                    {
                        singleUnderlyerIDRef_ = singleUnderlyerNode.Attributes["href"].Value;
                    }
                    else
                    {
                        singleUnderlyer_ = new SingleUnderlyer(singleUnderlyerNode);
                    }
                }
                else
                {
                    singleUnderlyer_ = new SingleUnderlyer(singleUnderlyerNode);
                }
            }
            
        
            XmlNode basketNode = xmlNode.SelectSingleNode("basket");
            
            if (basketNode != null)
            {
                if (basketNode.Attributes["href"] != null || basketNode.Attributes["id"] != null) 
                {
                    if (basketNode.Attributes["id"] != null) 
                    {
                        basketIDRef_ = basketNode.Attributes["id"].Value;
                        Basket ob = new Basket(basketNode);
                        IDManager.SetID(basketIDRef_, ob);
                    }
                    else if (basketNode.Attributes["href"] != null)
                    {
                        basketIDRef_ = basketNode.Attributes["href"].Value;
                    }
                    else
                    {
                        basket_ = new Basket(basketNode);
                    }
                }
                else
                {
                    basket_ = new Basket(basketNode);
                }
            }
            
        
        }
        
    
        #region SingleUnderlyer_
        private SingleUnderlyer singleUnderlyer_;
        public SingleUnderlyer SingleUnderlyer_
        {
            get
            {
                if (this.singleUnderlyer_ != null)
                {
                    return this.singleUnderlyer_; 
                }
                else if (this.singleUnderlyerIDRef_ != null)
                {
                    singleUnderlyer_ = IDManager.getID(singleUnderlyerIDRef_) as SingleUnderlyer;
                    return this.singleUnderlyer_; 
                }
                else
                {
                      return this.singleUnderlyer_; 
                }
            }
            set
            {
                if (this.singleUnderlyer_ != value)
                {
                    this.singleUnderlyer_ = value;
                }
            }
        }
        #endregion
        
        public string singleUnderlyerIDRef_ { get; set; }
        #region Basket_
        private Basket basket_;
        public Basket Basket_
        {
            get
            {
                if (this.basket_ != null)
                {
                    return this.basket_; 
                }
                else if (this.basketIDRef_ != null)
                {
                    basket_ = IDManager.getID(basketIDRef_) as Basket;
                    return this.basket_; 
                }
                else
                {
                      return this.basket_; 
                }
            }
            set
            {
                if (this.basket_ != value)
                {
                    this.basket_ = value;
                }
            }
        }
        #endregion
        
        public string basketIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-asset-5-4/Underlyer.cpp
// Underlyer.cpp 
#include "Underlyer.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

Underlyer::Underlyer(TiXmlNode* xmlNode)
: ISerialized(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //singleUnderlyerNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* singleUnderlyerNode = xmlNode->FirstChildElement("singleUnderlyer");

   if(singleUnderlyerNode){singleUnderlyerIsNull_ = false;}
   else{singleUnderlyerIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- singleUnderlyerNode , address : " << singleUnderlyerNode << std::endl;
   #endif
   if(singleUnderlyerNode)
   {
      if(singleUnderlyerNode->Attribute("href") || singleUnderlyerNode->Attribute("id"))
      {
          if(singleUnderlyerNode->Attribute("id"))
          {
             singleUnderlyerIDRef_ = singleUnderlyerNode->Attribute("id");
             singleUnderlyer_ = boost::shared_ptr<SingleUnderlyer>(new SingleUnderlyer(singleUnderlyerNode));
             singleUnderlyer_->setID(singleUnderlyerIDRef_);
             IDManager::instance().setID(singleUnderlyerIDRef_,singleUnderlyer_);
          }
          else if(singleUnderlyerNode->Attribute("href")) { singleUnderlyerIDRef_ = singleUnderlyerNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { singleUnderlyer_ = boost::shared_ptr<SingleUnderlyer>(new SingleUnderlyer(singleUnderlyerNode));}
   }

   //basketNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* basketNode = xmlNode->FirstChildElement("basket");

   if(basketNode){basketIsNull_ = false;}
   else{basketIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- basketNode , address : " << basketNode << std::endl;
   #endif
   if(basketNode)
   {
      if(basketNode->Attribute("href") || basketNode->Attribute("id"))
      {
          if(basketNode->Attribute("id"))
          {
             basketIDRef_ = basketNode->Attribute("id");
             basket_ = boost::shared_ptr<Basket>(new Basket(basketNode));
             basket_->setID(basketIDRef_);
             IDManager::instance().setID(basketIDRef_,basket_);
          }
          else if(basketNode->Attribute("href")) { basketIDRef_ = basketNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { basket_ = boost::shared_ptr<Basket>(new Basket(basketNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<SingleUnderlyer> Underlyer::getSingleUnderlyer()
{
   if(!this->singleUnderlyerIsNull_){
        if(singleUnderlyerIDRef_ != ""){
             return boost::shared_static_cast<SingleUnderlyer>(IDManager::instance().getID(singleUnderlyerIDRef_));
        }else{
             return this->singleUnderlyer_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SingleUnderlyer>();
   }
}
boost::shared_ptr<Basket> Underlyer::getBasket()
{
   if(!this->basketIsNull_){
        if(basketIDRef_ != ""){
             return boost::shared_static_cast<Basket>(IDManager::instance().getID(basketIDRef_));
        }else{
             return this->basket_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Basket>();
   }
}
}

